src/node/channels.h

```diff
@@ -246,93 +258,346 @@ namespace ccf
       }
 
-      auto network_pubk = crypto::make_public_key(network_kp->public_key_pem());
+      return true;
+    }
 
-      auto peer_public_size = serialized::read<size_t>(data, size);
-      auto peer_public_start = data;
+    bool verify_peer_signature(CBuffer msg, CBuffer sig)
+    {
+      LOG_DEBUG_FMT(
+        "Verifying peer signature with peer certificate serial {}",
+        peer_cv ? peer_cv->serial_number() : "no peer_cv!");
 
-      if (peer_public_size > size)
+      if (!peer_cv || !peer_cv->verify(msg, sig))
       {
         LOG_FAIL_FMT(
-          "Peer public key header wants {} bytes, but only {} remain",
-          peer_public_size,
-          size);
+          "Node channel peer signature verification failed for {} with "
+          "certificate serial {}",
+          peer_id,
+          peer_cv->serial_number());
         return false;
       }
 
-      data += peer_public_size;
-      size -= peer_public_size;
+      return true;
+    }
+
+    // Protocol overview:
+    //
+    // initiate()
+    // > key_exchange_init message
+    // consume_initiator_key_share() [by responder]
+    // < key_exchange_response message
+    // consume_responder_key_share() [by initiator]
+    // > key_exchange_final message
+    // check_peer_key_share_signature() [by responder]
+    // both reach status == ESTABLISHED
+
+    bool consume_responder_key_share(const std::vector<uint8_t>& data)
+    {
+      return consume_responder_key_share(data.data(), data.size());
+    }
 
-      auto signature_size = serialized::read<size_t>(data, size);
-      auto signature_start = data;
+    bool consume_responder_key_share(const uint8_t* data, size_t size)
+    {
+      LOG_DEBUG_FMT("status == {}", status);
 
-      if (signature_size > size)
+      if (status != INITIATED && status != ESTABLISHED)
       {
-        LOG_FAIL_FMT(
-          "Signature header wants {} bytes, but only {} remain",
-          signature_size,
-          size);
         return false;
       }
 
-      if (signature_size < size)
+      size_t peer_version = serialized::read<size_t>(data, size);
+      CBuffer ks = extract_buffer(data, size);
+      CBuffer sig = extract_buffer(data, size);
+      CBuffer pc = extract_buffer(data, size);
+
+      LOG_DEBUG_FMT(
+        "From responder {}: version={} ks={} sig={} pc={}",
+        peer_id,
+        peer_version,
+        ds::to_hex(ks),
+        ds::to_hex(sig),
+        ds::to_hex(pc));
+
+      if (size != 0)
       {
-        LOG_FAIL_FMT(
-          "Expected signature to use all remaining {} bytes, but only uses "
-          "{}",
-          size,
-          signature_size);
+        LOG_FAIL_FMT("{} exccess bytes remaining", size);
         return false;
       }
 
-      if (!network_pubk->verify(
-            peer_public_start,
-            peer_public_size,
-            signature_start,
-            signature_size))
+      if (peer_version != protocol_version || ks.n == 0 || sig.n == 0)
```

Comment:
It might be worth having a specific `LOG_FAIL` message if the protocols mismatch?